From e794e5db0490da2261aa360d47f86bbc836bde94 Mon Sep 17 00:00:00 2001
From: lisu <lisu@qti.qualcomm.com>
Date: Wed, 26 Nov 2025 22:08:09 -0800
Subject: [PATCH] kmssink: Do not close the DRM prime handle twice

In some cases, the GEM handles for multi planes may be the same. For
example, the NV12 surface may have the same GEM handle for the plane0
and the plane1. Then, the DRM_IOCTL_GEM_CLOSE will close the same handle
twice and get an "Invalid argument 22" error the second time. Check if
there are more than one planes using the same handle, do not close the
same handle twice.

Change-Id: Ie95e29917f244e5df25b3c79b29e3689b8eb0d4c
Signed-off-by: lisu <lisu@qti.qualcomm.com>
---
 sys/kms/gstkmsallocator.c     | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/sys/kms/gstkmsallocator.c b/sys/kms/gstkmsallocator.c
index 2ae22fcaec..4246992667 100644
--- a/sys/kms/gstkmsallocator.c
+++ b/sys/kms/gstkmsallocator.c
@@ -522,6 +522,7 @@ gst_kms_allocator_dmabuf_import (GstAllocator * allocator, gint * prime_fds,
   GstKMSMemory *kmsmem;
   GstMemory *mem;
   gint i, ret;
+  gint j;
 
   g_return_val_if_fail (n_planes <= GST_VIDEO_MAX_PLANES, FALSE);
 
@@ -548,11 +549,25 @@ gst_kms_allocator_dmabuf_import (GstAllocator * allocator, gint * prime_fds,
     struct drm_gem_close arg = { kmsmem->gem_handle[i], };
     gint err;
 
+    if (!kmsmem->gem_handle[i])
+      continue;
+
     err = drmIoctl (alloc->priv->fd, DRM_IOCTL_GEM_CLOSE, &arg);
     if (err)
       GST_WARNING_OBJECT (allocator,
           "Failed to close GEM handle: %s %d", g_strerror (errno), errno);
 
+    /* Check if there are any other planes using the same handle */
+    for (j = i + 1; j < n_planes; j++) {
+      if (kmsmem->gem_handle[j] == kmsmem->gem_handle[i]) {
+        GST_DEBUG_OBJECT (allocator,
+            "GEM handle %d corresponding to plane %d is already closed(duplicated to plane %d)",
+            kmsmem->gem_handle[j], j, i);
+        /* Do not close the same handle twice */
+        kmsmem->gem_handle[j] = 0;
+      }
+    }
+
     kmsmem->gem_handle[i] = 0;
   }
 
-- 
2.34.1

